Brave News Desktop Customize Toggle fixes

petemill opened the pull request:
Fix https://github.com/brave/brave-browser/issues/26519
- Desktop Customize UI Toggle will only disable show-on-NTP pref
- Desktop Customize UI updates data when enabled state changes
- Make sure any new code introduced with "v2" of Brave News does not contact

Fix https://github.com/brave/brave-browser/issues/26520
- Opt-in wording change and consistency for the Desktop Customize UI and the Desktop Opt-In card

Fix https://github.com/brave/brave-browser/issues/26678
- Add es_ES, es_MX and pt_BR to list of default-enabled locales

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
On issues

fallaciousreasoning commented on .storybook/locale.ts:
```diff
@@ -319,12 +319,6 @@ let locale: Record<string, string> = {
   showTopSites: 'Show Top Sites',
   showRewards: 'Show Rewards',
   tosAndPp: 'By turning on {{title}}, you agree to the $1Terms of Service$2 and $3Privacy Policy$4.',
-  braveTodayDisableSourceCommand: 'Disable content from $1',
```
:heart:

fallaciousreasoning commented on browser/ui/webui/brave_webui_source.cc:
```diff
@@ -189,11 +189,7 @@ void CustomizeWebUIHTMLSource(content::WebUI* web_ui,
         { "searchPromotionSearchBoxPlaceholderLabel", IDS_BRAVE_NEW_TAB_SEARCH_PROMOTION_SEARCH_BOX_PLACEHOLDER},  // NOLINT
 
         { "braveTodayTitle", IDS_BRAVE_TODAY_TITLE },
-        { "braveTodayIntroTitle", IDS_BRAVE_TODAY_INTRO_TITLE },
```
:heart:

fallaciousreasoning commented on components/brave_new_tab_ui/components/default/braveToday/cardIntro.ts:
```diff
@@ -13,26 +13,34 @@ export const Intro = styled(StandardBlock)`
   line-height: 17px;
   letter-spacing: 0.16px;
+
+  & + & {
```
Stylish!

fallaciousreasoning commented on components/brave_new_tab_ui/components/default/braveToday/cards/cardNoContent.tsx:
```diff
@@ -1,7 +1,7 @@
 // This Source Code Form is subject to the terms of the Mozilla Public
 // License, v. 2.0. If a copy of the MPL was not distributed with this file,
-// you can obtain one at http://mozilla.org/MPL/2.0/.
+// you can obtain one at https://mozilla.org/MPL/2.0/.
```
:heart:

fallaciousreasoning commented on components/brave_new_tab_ui/components/default/braveToday/cards/cardOptIn.tsx:
```diff
@@ -14,27 +14,31 @@ type Props = {
+        </Card.Paragraph>
+        <Card.Paragraph>
+          {descriptionTwoTextParts.beforeTag}
+          <a href={'https://brave.com/privacy/browser/'}>
```
Not sure what our style guide says about `{}` on urls?

```suggestion
          <a href='https://brave.com/privacy/browser/'>
```

petemill replied:
Our linter used to enforce props should be surrounded by {} - I thought for all JSX elements, but it doesn't seem to be enforced either way atm. I'm not sure where the team wants to fall on this.

fallaciousreasoning commented on components/brave_new_tab_ui/components/default/braveToday/customize/Context.tsx:
```diff
@@ -29,6 +30,10 @@ interface BraveNewsContext {
   // Publishers to suggest to the user.
   suggestedPublisherIds: string[]
   updateSuggestedPublisherIds: () => void
+  optInPrefEnabled: boolean | undefined
```
I don't know how I feel about exposing these in multiple places, rather than just going through the prefs hook.

Maybe instead of just reexposing the `useNewTabPref` stuff we could have a function which does both setOptIn & setShow?

const setShowBraveNews(optIn: boolean, show: boolean) => { ... }

That way we couldn't forget to do one of them (which was the cause of @simonhong's issue the other day).

fallaciousreasoning replied:
If not, I think I'd prefer users of the pref to just go straight through the `useNewTabPref` rather than drilling it through the Context here.

fallaciousreasoning commented on components/brave_new_tab_ui/components/default/braveToday/customize/Context.tsx:
```diff
@@ -40,14 +45,26 @@ export const BraveNewsContext = React.createContext<BraveNewsContext>({
+
+  // Update initially and when opt-in / enabled changes
+  React.useEffect(() => {
+    api.update()
```
:+1: that's elegant!

fallaciousreasoning commented on components/brave_new_tab_ui/components/default/braveToday/customize/Configure.tsx:
```diff
@@ -109,12 +108,29 @@ const Content = styled.div`
+      setShowOnNTPPrefEnabled(true)
+      return
+    }
+    setShowOnNTPPrefEnabled(false)
```
So there's no way to opt out?

petemill replied:
correct. The opt-in just means that you've seen the terms, really. You can still turn the feature off which will disable all the fetching.

fallaciousreasoning commented on components/brave_new_tab_ui/components/default/braveToday/customize/Configure.tsx:
```diff
@@ -109,12 +108,29 @@ const Content = styled.div`
+    showOnNTPPrefEnabled
+  } = useBraveNews()
+
+  const handleEnabledChange = (shouldEnable: boolean) => {
```
Maybe we should put this function on the `Context`?

fallaciousreasoning commented on components/brave_new_tab_ui/components/default/braveToday/customize/Configure.tsx:
```diff
@@ -109,12 +108,29 @@ const Content = styled.div`
+    setShowOnNTPPrefEnabled(false)
+  }
+
+  const isBraveNewsFullyEnabled = optInPrefEnabled && showOnNTPPrefEnabled
```
And this flag, rather than the two individual ones?

fallaciousreasoning commented on components/brave_new_tab_ui/stories/today.tsx:
```diff
@@ -98,6 +100,10 @@ export const Error = () => (
 )
 
+export const OptIn = () => (
+  <BraveTodayOptInCard onOptIn={() => console.log('opt-in clicked')} onDisable={() => console.log('disable clicked')} />
```
Do we want to remove the logs?

petemill replied:
I think it's fine for the story

fallaciousreasoning commented on components/brave_new_tab_ui/stories/todayStrings.ts:
```diff
@@ -0,0 +1,17 @@
+import { provideStrings } from '../../../.storybook/locale'
+
+provideStrings({
+  braveTodayDisableSourceCommand: 'Disable content from $1',
```
Oh cool, I didn't know we could do this!

fallaciousreasoning reviewed (approved):
LGTM, a bunch of comments but most of them are just chatty

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://brave-jenkins-build-artifacts.s3.amazonaws.com/brave-browser-build-pr/news-v2-toggle/d5b14a213811b87dbc7378ff3179e68bd9f4c7b8/storybook/index.html)

fallaciousreasoning reviewed (approved):
Brave News: Add es_ES es_MX and pt_BR to default shown on NTP LGTM

petemill commented:
I addressed some of your feedback @fallaciousreasoning to move some things around on the Context. As discussed unfortunately I think we'll have to keep useNewTabPref on the Context for now and remove when we move Brave News hooks to a shared component.

fallaciousreasoning reviewed (approved):
Still LGTM

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://brave-jenkins-build-artifacts.s3.amazonaws.com/brave-browser-build-pr/news-v2-toggle/7e14a5eeea03669c8a0e91b6edf06b852a10d215/storybook/index.html)

kjozwiak commented:
Verification was completed on `master` via https://github.com/brave/brave-core/pull/15897#issuecomment-1311314001 so we can uplift https://github.com/brave/brave-core/pull/15897 into `1.46.x`.
